File: ios/Classes/nativeModel/PrinterInfo.swift
//
//  PrinterInfo.swift
//  zebrautil
//
//  Native model for printer information to ensure type safety
//  between Dart and native communication
//

import Foundation

@objc public class PrinterInfo: NSObject {
    // Required properties
    @objc public let address: String
    @objc public let name: String
    @objc public let status: String
    @objc public let isWifi: Bool
    @objc public let isBluetooth: Bool
    @objc public let connectionType: String
    @objc public let discoveryMethod: String
    
    // Optional properties
    @objc public let model: String?
    @objc public let manufacturer: String?
    @objc public let firmwareRevision: String?
    @objc public let hardwareRevision: String?
    @objc public let displayName: String?
    @objc public let port: Int
    @objc public let dnsName: String?
    
    @objc public init(
        address: String,
        name: String,
        status: String = "Available",
        isWifi: Bool = false,
        isBluetooth: Bool = false,
        connectionType: String = "Unknown",
        discoveryMethod: String = "unknown",
        model: String? = nil,
        manufacturer: String? = "Zebra",
        firmwareRevision: String? = nil,
        hardwareRevision: String? = nil,
        displayName: String? = nil,
        port: Int = 9100,
        dnsName: String? = nil
    ) {
        self.address = address
        self.name = name
        self.status = status
        self.isWifi = isWifi
        self.isBluetooth = isBluetooth
        self.connectionType = connectionType
        self.discoveryMethod = discoveryMethod
        self.model = model
        self.manufacturer = manufacturer
        self.firmwareRevision = firmwareRevision
        self.hardwareRevision = hardwareRevision
        self.displayName = displayName
        self.port = port
        self.dnsName = dnsName
        
        super.init()
    }
    
    /// Convert to dictionary for channel communication
    @objc public func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "Address": address,
            "Name": name,
            "Status": status,
            "IsWifi": isWifi,
            "isBluetooth": isBluetooth,
            "connectionType": connectionType,
            "discoveryMethod": discoveryMethod,
            "port": port,
            "brand": manufacturer ?? "Zebra"
        ]
        
        // Add optional properties if present
        if let model = model {
            dict["model"] = model
        }
        if let manufacturer = manufacturer {
            dict["manufacturer"] = manufacturer
        }
        if let firmwareRevision = firmwareRevision {
            dict["firmwareRevision"] = firmwareRevision
        }
        if let hardwareRevision = hardwareRevision {
            dict["hardwareRevision"] = hardwareRevision
        }
        if let displayName = displayName {
            dict["displayName"] = displayName
        }
        if let dnsName = dnsName {
            dict["dnsName"] = dnsName
        }
        
        return dict
    }
    
    /// Create from dictionary received from channel
    @objc public static func fromDictionary(_ dict: [String: Any]) -> PrinterInfo? {
        guard let address = dict["Address"] as? String else { return nil }
        
        return PrinterInfo(
            address: address,
            name: dict["Name"] as? String ?? "Unknown Printer",
            status: dict["Status"] as? String ?? "Available",
            isWifi: dict["IsWifi"] as? Bool ?? false,
            isBluetooth: dict["isBluetooth"] as? Bool ?? false,
            connectionType: dict["connectionType"] as? String ?? "Unknown",
            discoveryMethod: dict["discoveryMethod"] as? String ?? "unknown",
            model: dict["model"] as? String,
            manufacturer: dict["manufacturer"] as? String ?? "Zebra",
            firmwareRevision: dict["firmwareRevision"] as? String,
            hardwareRevision: dict["hardwareRevision"] as? String,
            displayName: dict["displayName"] as? String,
            port: dict["port"] as? Int ?? 9100,
            dnsName: dict["dnsName"] as? String
        )
    }
}

File: lib/internal/native_models/printer_info.dart
/// Native model for printer information to ensure type safety
/// between Dart and native communication
library;

import '../../models/zebra_device.dart';

class NativePrinterInfo {
  const NativePrinterInfo({
    required this.address,
    required this.name,
    this.status = 'Available',
    this.isWifi = false,
    this.isBluetooth = false,
    this.connectionType = 'Unknown',
    this.discoveryMethod = 'unknown',
    this.model,
    this.manufacturer = 'Zebra',
    this.firmwareRevision,
    this.hardwareRevision,
    this.displayName,
    this.port = 9100,
    this.dnsName,
  });

  /// Create from native dictionary
  factory NativePrinterInfo.fromNative(Map<String, dynamic> data) {
    return NativePrinterInfo(
      address: data['Address'] as String? ?? '',
      name: data['Name'] as String? ?? 'Unknown Printer',
      status: data['Status'] as String? ?? 'Available',
      isWifi: data['IsWifi'] == true || data['IsWifi'] == 'true',
      isBluetooth: data['isBluetooth'] == true || data['isBluetooth'] == 'true',
      connectionType: data['connectionType'] as String? ?? 'Unknown',
      discoveryMethod: data['discoveryMethod'] as String? ?? 'unknown',
      model: data['model'] as String?,
      manufacturer: data['manufacturer'] as String? ?? 'Zebra',
      firmwareRevision: data['firmwareRevision'] as String?,
      hardwareRevision: data['hardwareRevision'] as String?,
      displayName: data['displayName'] as String?,
      port: data['port'] as int? ?? 9100,
      dnsName: data['dnsName'] as String?,
    );
  }

  final String address;
  final String name;
  final String status;
  final bool isWifi;
  final bool isBluetooth;
  final String connectionType;
  final String discoveryMethod;

  // Optional properties
  final String? model;
  final String? manufacturer;
  final String? firmwareRevision;
  final String? hardwareRevision;
  final String? displayName;
  final int port;
  final String? dnsName;

  /// Convert to native dictionary format
  Map<String, dynamic> toNative() {
    final Map<String, dynamic> dict = {
      'Address': address,
      'Name': name,
      'Status': status,
      'IsWifi': isWifi,
      'isBluetooth': isBluetooth,
      'connectionType': connectionType,
      'discoveryMethod': discoveryMethod,
      'port': port,
    };

    // Add optional properties if present
    if (model != null) dict['model'] = model;
    if (manufacturer != null) dict['manufacturer'] = manufacturer;
    if (firmwareRevision != null) dict['firmwareRevision'] = firmwareRevision;
    if (hardwareRevision != null) dict['hardwareRevision'] = hardwareRevision;
    if (displayName != null) dict['displayName'] = displayName;
    if (dnsName != null) dict['dnsName'] = dnsName;

    return dict;
  }

  /// Convert to ZebraDevice model
  ZebraDevice toZebraDevice() {
    return ZebraDevice(
      address: address,
      name: name,
      status: status,
      isWifi: isWifi,
      isBluetooth: isBluetooth,
      brand: manufacturer ?? 'Zebra',
      model: model,
      displayName: displayName ?? _generateDisplayName(),
      manufacturer: manufacturer ?? 'Zebra',
      firmwareRevision: firmwareRevision,
      hardwareRevision: hardwareRevision,
      connectionType: connectionType,
      port: port,
    );
  }

  String _generateDisplayName() {
    if (model != null && model!.isNotEmpty) {
      return 'Zebra $model - $name';
    } else {
      return 'Zebra Printer - $name';
    }
  }

  @override
  String toString() {
    return 'NativePrinterInfo(address: $address, name: $name, status: $status, '
        'isWifi: $isWifi, isBluetooth: $isBluetooth, connectionType: $connectionType, '
        'discoveryMethod: $discoveryMethod, model: $model, manufacturer: $manufacturer, '
        'port: $port)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is NativePrinterInfo &&
        other.address == address &&
        other.port == port;
  }

  @override
  int get hashCode => address.hashCode ^ port.hashCode;
}